### PR TITLE
[MNG-8024] Make WrapperProperties and WrapperList serizalizable.

### PR DESCRIPTION
**Pull request by cstamas**
This solves the two, but there are more...

---

https://issues.apache.org/jira/browse/MNG-8024

### CODE REVIEW

**Comment by gnodet** [comment]
We need to extend the unit test https://github.com/apache/maven/blob/master/maven-model/src/test/java/org/apache/maven/model/SerializationTest.java to exercise the actual problem.

File: src/mdo/java/WrapperList.java [hunk 1]
```diff
@@ -27,16 +27,21 @@
 import java.util.function.Supplier;
 
 class WrapperList<T, U> extends AbstractList<T> {
```
Comment (gnodet) [inline src/mdo/java/WrapperList.java:29]: What's the point is the `WrapperList` itself does not implement `Serializable` ?

**Review by gnodet — CHANGES_REQUESTED** [review]
We need a clear unit test.  Basic serialisation is already supported, so something's wrong...

**Comment by gnodet** [comment]
> We need a clear unit test. Basic serialisation is already supported, so something's wrong... The `WrapperList` and `WrapperProperties` should not actually be used during serialisation, as those objects are not present in fields.

I wonder if the problem is when serializing part of the model instead of the whole `Model`.

**Comment by cstamas** [comment]
Yes, that is exactly what is happening, the incremental build basically taks "bits" of model and would like to serialize those with DataOutputStream...

**Comment by gnodet** [comment]
> Yes, that is exactly what is happening, the incremental build basically taks "bits" of model and would like to serialize those with DataOutputStream...

I'm not sure we should support this use case, as I think the whole model will be serialised anyway.  The lambdas in the list/properties objects do retain a reference to the parent up to the Model.  This is needed in order to implement the modification methods while keeping the original reference.

So while this may actually make things work, I'm not completely sure we should do it...  It may be easier to simplify check if there's a `getDelegate()` method, and if so, serialize that object instead.

**Comment by gnodet** [comment]
> > Yes, that is exactly what is happening, the incremental build basically taks "bits" of model and would like to serialize those with DataOutputStream...
> 
> I'm not sure we should support this use case, as I think the whole model will be serialised anyway. The lambdas in the list/properties objects do retain a reference to the parent up to the Model. This is needed in order to implement the modification methods while keeping the original reference.
> 
> So while this may actually make things work, I'm not completely sure we should do it... It may be easier to simplify check if there's a `getDelegate()` method, and if so, serialize that object instead.

Another way would be to generate writeObject / readObject methods that will actually delegate to... the delegates from the v4 model.

**Pull request closed without merging**